### ISSUE TITLE
Type inference of methods of Pytorch's DataParallel class

### ISSUE THREAD

**Issue by tianze0926** [microsoft/pyright#4776]
**Describe the bug**

In the following code, the type of  variable `r` was inferred as `Tensor | Module` instead of a method as it should be according to the [source code](https://pytorch.org/docs/stable/_modules/torch/nn/parallel/data_parallel.html#DataParallel).

```python
from torch.nn import DataParallel
d = DataParallel()
r = d.replicate
reveal_type(r)
```

**To Reproduce**
Environment: conda
- python: 3.10.9
- pytorch: 2.0.0

Run `pyright $FILE.py`.

**Expected behavior**
The type of `r` should be a method.

**VS Code extension or command-line**
pyright command-line 1.1.299

**Comment by tianze0926** [comment]
@erictraut Thanks for your reply. But forgive me if I missed something since I am new to Python's typing, what do you mean by there is no defined "attribute"? Because I do see the method `replicate` is defined in the [source code](https://pytorch.org/docs/stable/_modules/torch/nn/parallel/data_parallel.html#DataParallel). Does it have to be declared in `__getattr__` of `DataParallel`?

**Comment by tianze0926** [comment]
The method seems to exist in my installation:
```
>>> from torch.nn import DataParallel
>>> from torch.nn.modules import Conv2d
>>> d = DataParallel(Conv2d(3, 3, 3))
>>> r = d.replicate
>>> type(r)
<class 'method'>
>>> r
<bound method DataParallel.replicate of DataParallel(
  (module): Conv2d(3, 3, kernel_size=(3, 3), stride=(1, 1))
)>
```

When I jump to the source code on VS Code, the `replicate` method can also be found:
![image](https://user-images.githubusercontent.com/66857124/225337572-9e9a4861-e831-4ec6-ada1-c77e8e1b1937.png)

But pyright gives the following output:
```
No configuration file found.
No pyproject.toml file found.
stubPath /ceph/home/zhoutz19/test/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
pyright 1.1.299
****/class_typing.py
  ****/class_typing.py:6:13 - information: Type of "r" is "Tensor | Module"
0 errors, 0 warnings, 1 information 
Completed in 1.518sec
```

Therefore when I call this method Pyright would generate `error: Object of type "Tensor" is not callable (reportGeneralTypeIssues)`.  So I am very confused and not sure what's the cause.

**Comment by erictraut** [comment]
You are looking at the file `data_parallel.py`. Torch 1.13.1 also includes a type stub file `data_parallel.pyi`. Type checkers like pyright always prefer type stubs (".pyi" files) over source (".py") files if both are present. In this case, it appears that the type stub file is inconsistent with the implementation of the `DataParallel` class, so this is a bug in the type stub file provided as part of the torch distribution. Please report it to the maintainers of torch.

**Comment by tianze0926** [comment]
Oh I see. Thank you very much!
